### PR TITLE
mixtral: drop training-branching hack for SFT segfault & add ZeRO-3 leaf utility

### PR DESCRIPTION
**Pull request by yafshar**
# What does this PR do?

### 1. Removes the temporary MOE-kernel workaround
- Deletes the TODO block `if self.training … else …` branch that selected different HPU MOE kernels.  
- Replaces both `call_sparse_moe_op` and `call_dynamic_moe_op` with the single HPU-optimized
  `torch.ops.hpu.mixture_of_experts` after Synapse `1.21.0` fixed the segfault reported in PR #1798.  
- Keeps the post-kernel all-reduce for inference unchanged.

### 2. Adds reusable ZeRO-3 leaf-promotion utility
- Introduces `optimum/habana/distributed/zero3_utils.py` containing
  `apply_zero3_leaf_promotion(model)`. 
(See https://github.com/deepspeedai/DeepSpeed/blob/master/deepspeed/utils/z3_leaf_module.py#L70)
- Exports the helper from `optimum.habana.distributed` so any training script must check if `deepspeed` is imported and ZeRO-3 is active to call it.

### 3. Wires the utility into `sft.py` script
- Update the README

### 4. Provides new ZeRO-3 config template for Mixtral
- Adds `examples/language-modeling/mixtral_zero3_config.json`
- This config enables ZeRO Stage 3 with overlap communication to support `torch.ops.hpu.mixture_of_experts`

Tests:

main
```sh
>>> PT_HPU_LAZY_MODE=1 PT_ENABLE_INT64_SUPPORT=1 python ../gaudi_spawn.py --world_size 4 --use_deepspeed sft.py \
--model_name_or_path mistralai/Mixtral-8x7B-Instruct-v0.1 \
--dataset_name "philschmid/dolly-15k-oai-style" \
--subset 'data/' \
--streaming False \
--deepspeed ../language-modeling/llama2_ds_zero3_config.json \
--output_dir="./model_mixtral" \
--do_train \
--max_steps=500 \
--logging_steps=10 \
--save_steps=100 \
--per_device_train_batch_size=2 \
--per_device_eval_batch_size=1 \
--gradient_accumulation_steps=2 \
--learning_rate=1e-4 \
--lr_scheduler_type="cosine" \
--warmup_steps=100 \
--weight_decay=0.05 \
--optim="paged_adamw_32bit" \
--lora_target_modules "q_proj" "v_proj" \
--bf16 \
--remove_unused_columns=False \
--max_seq_length 512 \
--run_name="sft_mixtral" \
--report_to=none \
--use_habana \
--use_lazy_mode


***** train metrics *****
  epoch                       =     1.3972
  max_memory_allocated (GB)   =      80.43
  memory_allocated (GB)       =      23.33
  total_flos                  =   107879GF
  total_memory_available (GB) =     126.54
  train_loss                  =     5.1121
  train_runtime               = 0:33:59.01
  train_samples_per_second    =      3.923
  train_steps_per_second      =      0.245
```

this PR
```sh
>>> PT_HPU_LAZY_MODE=1 PT_ENABLE_INT64_SUPPORT=1 python ../gaudi_spawn.py --world_size 4 --use_deepspeed sft.py \
--model_name_or_path mistralai/Mixtral-8x7B-Instruct-v0.1 \
--dataset_name "philschmid/dolly-15k-oai-style" \
--subset 'data/' \
--streaming False \
--deepspeed ../language-modeling/mixtral_ds_zero3_config.json \
--output_dir="./model_mixtral" \
--do_train \
--max_steps=500 \
--logging_steps=10 \
--save_steps=100 \
--per_device_train_batch_size=2 \
--per_device_eval_batch_size=1 \
--gradient_accumulation_steps=2 \
--learning_rate=1e-4 \
--lr_scheduler_type="cosine" \
--warmup_steps=100 \
--weight_decay=0.05 \
--optim="paged_adamw_32bit" \
--lora_target_modules "q_proj" "v_proj" \
--bf16 \
--remove_unused_columns=False \
--max_seq_length 512 \
--run_name="sft_mixtral" \
--report_to=none \
--use_habana \
--use_lazy_mode \
--use_zero3_leaf_promotion

***** train metrics *****
  epoch                       =     1.3972
  max_memory_allocated (GB)   =      46.23
  memory_allocated (GB)       =      23.49
  total_flos                  =   107879GF
  total_memory_available (GB) =     126.54
  train_loss                  =     5.1126
  train_runtime               = 0:21:16.90
  train_samples_per_second    =      6.265
  train_steps_per_second      =      0.392
```

### 📊 Training Performance Comparison

| Metric                        | Main                  | This PR                 | Improvement / Change        |
|------------------------------|----------------------------|----------------------------|-----------------------------|
| **Train Runtime**            | 33 min 59 sec              | 21 min 16 sec              | ⬇️ ~37.4% faster             |
| **Train Samples/sec**        | 3.923                      | 6.265                      | ⬆️ ~59.6% increase           |
| **Train Steps/sec**          | 0.245                      | 0.392                      | ⬆️ ~60% increase             |
| **Max Memory Allocated (GB)**| 80.43                      | 46.23                      | ⬇️ ~42.5% less memory usage  |
| **Train Loss**               | 5.1121                     | 5.1126                     | ⬆️ Negligible increase       |

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
> Nice PR! I think it's worth adding a regression test in https://github.com/huggingface/optimum-habana/blob/main/tests/test_examples.py. You can use the same command you provided in this PR.

Thanks! It takes about 30 minutes to run, which is why I initially left it out. Please let me know if you'd like me to include it.

**Comment by regisss** [comment]
> > Nice PR! I think it's worth adding a regression test in https://github.com/huggingface/optimum-habana/blob/main/tests/test_examples.py. You can use the same command you provided in this PR.
> 
> Thanks! It takes about 30 minutes to run, which is why I initially left it out. Please let me know if you'd like me to include it.

I think it's okay to include it. Worst case, I'll make it run less training steps later.

**Comment by yafshar** [comment]
@regisss I added the test, just need to double check the reference numbers and then I will ping you

**Comment by yafshar** [comment]
@regisss The PR is ready for your review. test commands are updated for 8 HPUs, so those and reference numbers can be further optimized in the future. For now, I followed the steps outlined in the README to mimic the test setup. I also excluded perplexity due to the long runtime, but it can be added later if needed.

File: examples/trl/sft.py [hunk 1]
```diff
@@ -142,11 +143,13 @@ def create_datasets(tokenizer, args, seed=None):
         return train_data, valid_data, formating_func
 
     low_cpu_mem_usage = True
+    is_deepspeed_zero3_enabled = False
     if is_deepspeed_available():
-        from transformers.integrations.deepspeed import is_deepspeed_zero3_enabled
+        from transformers.integrations.deepspeed import is_deepspeed_zero3_enabled as ds_zero3_flag
 
-        if is_deepspeed_zero3_enabled():
+        if ds_zero3_flag():
```
Comment (IlyasMoutawwakil) [inline examples/trl/sft.py]: no need for the rename, ds_zero3_flag could be confusing and is_deepspeed_zero3_enabled  is more widely used in the lib https://github.com/search?q=repo%3Ahuggingface%2Foptimum-habana%20is_deepspeed_zero3_enabled%20&type=code

File: examples/trl/sft.py [hunk 2]
```diff
@@ -164,6 +167,12 @@ def create_datasets(tokenizer, args, seed=None):
     base_model.generation_config.flash_attention_recompute = script_args.flash_attention_recompute
     base_model.generation_config.flash_attention_causal_mask = script_args.flash_attention_causal_mask
 
+    apply_zero3_leaf_promotion(
+        base_model,
+        is_deepspeed_zero3_enabled=is_deepspeed_zero3_enabled,
+        use_zero3_leaf_promotion=training_args.use_zero3_leaf_promotion,
+    )
```
Comment (IlyasMoutawwakil) [inline examples/trl/sft.py]: I'd say no need for `is_deepspeed_zero3_enabled = True` and simply gard here on is_deepspeed_available() and is_deepspeed_zero3_enabled()

File: optimum/habana/distributed/zero3_utils.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import torch
+
+
+logger = logging.getLogger(__name__)
+
+_MODEL_Z3_LEAF_REGISTRY: Dict[str, Tuple[str, str]] = {
+    "llama": ("optimum.habana.transformers.models.llama.modeling_llama", "GaudiLlamaDecoderLayer"),
+    "mixtral": ("optimum.habana.transformers.models.mixtral.modeling_mixtral", "GaudiMixtralSparseMoeBlock"),
+    "qwen3_moe": ("optimum.habana.transformers.models.qwen3_moe.modeling_qwen3_moe", "GaudiQwen3MoeSparseMoeBlock"),
```
Comment (IlyasMoutawwakil) [inline optimum/habana/distributed/zero3_utils.py]: is there a specific reason why these classes are not explicitly imported here ? the `importlib.import_module` has tendency of making import errors hard to debug, so I think it should only be used when necessary (if importing the class triggers something for example)

File: tests/test_examples.py [hunk 4]
```diff
@@ -709,11 +719,13 @@ def _create_command_line(
...
                 f"--per_device_train_batch_size {train_batch_size}",
                 f"--per_device_eval_batch_size {eval_batch_size}",
-                f" --num_train_epochs {num_epochs}",
+                f"--num_train_epochs {num_epochs}",
                 "--use_habana",
                 "--throughput_warmup_steps 3",
                 "--save_strategy no",
             ]
+            if not gaudi_config_name:
+                cmd_line.pop(cmd_line.index(f"--gaudi_config_name {gaudi_config_name}"))
```
Comment (IlyasMoutawwakil) [inline tests/test_examples.py]: you mean if not None ?

Comment (yafshar) [reply tests/test_examples.py]: No, `gaudi_config_name` in this case for mixtral is an empty string "", please look at `test/utils.py`

File: optimum/habana/distributed/zero3_utils.py (hunk 3 above)
Comment (yafshar) [reply optimum/habana/distributed/zero3_utils.py]: The use of `importlib.import_module` here is intentional to avoid importing all model-specific modules upfront. That said, I agree that dynamic imports can obscure errors and make debugging harder.  But the registry is expected to grow (e.g., deepseek_v2, deepseek_v3, llama4, etc.), so I opted for a lazy approach. That said, if you prefer explicit imports here, I’m happy to update the code accordingly.

File: tests/test_examples.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply tests/test_examples.py:728]: it still doesn't make sense, the expression only valuates to True if gaudi_config_name is "" ? 😅

Comment (IlyasMoutawwakil) [reply tests/test_examples.py]: okay I see, but let's not play with truthiness, It makes more sense if `"--gaudi_config_name {gaudi_config_name}"` is never added to cmd_line when gaudi_config_name is an empty string (instead of adding it and then removing it)

File: examples/trl/sft.py [hunk 5]
```diff
@@ -164,6 +167,12 @@ def create_datasets(tokenizer, args, seed=None):
     base_model.generation_config.flash_attention_recompute = script_args.flash_attention_recompute
     base_model.generation_config.flash_attention_causal_mask = script_args.flash_attention_causal_mask
 
+    apply_zero3_leaf_promotion(
+        base_model,
+        is_deepspeed_zero3_enabled=is_zero3_enabled,
+        use_zero3_leaf_promotion=training_args.use_zero3_leaf_promotion,
+    )
```
Comment (IlyasMoutawwakil) [inline examples/trl/sft.py]: what do you think of `if is_zero3_enabled and training_args.use_zero3_leaf_promotion: apply_zero3_leaf_promotion(base_model)`
where the method apply_zero3_leaf_promotion is only called when it's functionality is needed

File: examples/trl/README.md [hunk 6]
```diff
@@ -116,7 +116,7 @@ PT_HPU_MAX_COMPOUND_OP_SIZE=10 PT_HPU_LAZY_MODE=1 python3 ../gaudi_spawn.py --wo
         --dataset_name "philschmid/dolly-15k-oai-style" \
         --subset 'data/' \
         --streaming False \
-        --deepspeed ../language-modeling/llama2_ds_zero3_config.json \
+        --deepspeed ./mixtral_ds_zero3_config.json \
```
Comment (IlyasMoutawwakil) [inline examples/trl/README.md]: should we not move this config to ./language-modeling ?

**Review by IlyasMoutawwakil — APPROVED** [review]
LGTM ! thanks for iterating on this ! I left one last nit but it's not important

File: examples/trl/README.md (hunk 6 above)
Comment (yafshar) [reply examples/trl/README.md]: I am not sure, and I am open to it. @regisss WDYT?

Comment (IlyasMoutawwakil) [reply examples/trl/README.md]: let's move it 🤗 better than having it next to the scripts https://github.com/huggingface/optimum-habana/tree/6b158602efb96a1efa54b71abf6f7b265f1b8ea7/examples/trl

Comment (yafshar) [reply examples/trl/README.md]: done!

Comment (regisss) [reply examples/trl/README.md]: Yes sounds good!

**Review by regisss — APPROVED** [review]
LGTM :rocket:
